src/dotnet/App.Maui/Platforms/Android/CronetMessageHandler.cs

```diff
@@ -0,0 +1,190 @@
+                _responseBodyChannel?.DisposeSilently();
+                _responseBodyChannel = null;
+                _responseBody = null;
+                p0.Cancel();
+            }
+            p0.FollowRedirect();
+        }
+
+        public override void OnResponseStarted(UrlRequest p0, UrlResponseInfo p1)
+        {
+            if (CancellationToken.IsCancellationRequested) {
+                p0.Cancel();
+                return;
+            }
+            _responseBody = MemoryStreamManager.GetStream();
+            _responseBodyChannel = Java.Nio.Channels.Channels.NewWritableChannel(_responseBody)!;
+
+            p0.Read(ByteBuffer.AllocateDirect(4096));
+        }
+
+        public override void OnReadCompleted(UrlRequest p0, UrlResponseInfo p1, ByteBuffer p2)
+        {
+            if (CancellationToken.IsCancellationRequested) {
+                p0.Cancel();
+                return;
+            }
+            p2.Flip();
+            try
+            {
+                _responseBodyChannel ??= Java.Nio.Channels.Channels.NewWritableChannel(_responseBody)!;
+                _responseBodyChannel.Write(p2);
+                p2.Clear();
+                p0.Read(p2);
+            }
+            catch (Exception e)
+            {
+                p0.Cancel();
+                _result.TrySetException(e);
+            }
+        }
+
+        public override void OnSucceeded(UrlRequest p0, UrlResponseInfo p1)
+        {
+            _responseBodyChannel?.DisposeSilently();
+            _responseBodyChannel = null;
+
+            var ret = new HttpResponseMessage ((HttpStatusCode)p1.HttpStatusCode) {
+                RequestMessage = RequestMessage,
+                ReasonPhrase = p1.HttpStatusText,
+                Version = string.IsNullOrWhiteSpace(p1.NegotiatedProtocol)
+                    ? HttpVersion.Version11
+                    : p1.NegotiatedProtocol == "h2"
+                        ? HttpVersion.Version20
+                        : HttpVersion.Version30,
+            };
+            if (_responseBody != null) {
+                _responseBody.Flush();
+                _responseBody.Position = 0;
+                ret.Content = new StreamContent(_responseBody);
+                _responseBody = null;
+            }
+            else
+                ret.Content = new StreamContent(Stream.Null);
+
+            foreach (var (key, values) in p1.AllHeaders)
+                if (key.StartsWith("content",StringComparison.OrdinalIgnoreCase))
+                    ret.Content.Headers.Add(key,values);
+                else
+                    ret.Headers.Add(key, values);
+            // we are waiting for transmit completion - but we can try to return Response as soon as possible with uncompleted StreamContent
+            _result.TrySetResult(ret);
+        }
+
+        public override void OnCanceled(UrlRequest request, UrlResponseInfo info)
+        {
+            _responseBodyChannel?.DisposeSilently();
```

Comment:
Prob. makes sense to move all the disposal code to a single helper method (seems it's repeating in a few places).